[16.0][IMP] sale_order_product_picker: Support use of discounts

CarlosRoca13 opened the pull request:
cc @Tecnativa TT48323

ping @sergio-teruel @carlosdauden

carlosdauden committed:
[IMP] sale_order_product_picker: Improve parameter options in readme

CarlosRoca13 committed:
[IMP] sale_order_product_picker: Support use of discounts

sergio-teruel commented on sale_order_product_picker/models/sale_order_picker.py:
```diff
@@ -100,21 +287,32 @@ def _get_last_sale_price_product(self, use_delivery_address=False):
         self.ensure_one()
         so_line = self.order_id.order_line.new({"product_id": self.product_id.id})
-        so_line.product_id_change()
+        # so_line._product_id_change()
```
Is ok this commented line??

CarlosRoca13 replied:
The method doesn't exist now, removed line

chienandalu commented on sale_order_product_picker/static/src/x2many/x2many_field.esm.js:
```diff
@@ -45,16 +72,21 @@ patch(X2ManyField.prototype, "sale_order_product_picker.X2ManyField", {
+            const newUpdateRecord = async (record) => {
                 // Function used when updating a record.
-                const res = updateRecord(record);
+                lineList = this.props.record.model.root.data.order_line;
```
You aren't declaring it... doesn't this fail?

chienandalu commented on sale_order_product_picker/static/src/x2many/x2many_field.esm.js:
```diff
@@ -67,28 +99,41 @@ patch(X2ManyField.prototype, "sale_order_product_picker.X2ManyField", {
+            const newRemoveRecord = async (record) => {
                 // Function used when removing a record.
-                const res = removeRecord(record);
+                lineList = this.props.record.model.root.data.order_line;
```
Same here..

CarlosRoca13 commented:
That var is defined in line 29. I added comments explaining why need to redefine the list.

CarlosRoca13 committed:
[IMP] sale_order_product_picker: Changes to improve usability of +1

Added changes to make the button +1 faster. With these changes, what we
do is to store the changes to be made while pressing the +1 button. Each
time the button is pressed, the timer that will execute the writing of
the lines is reset. The time to write the pending lines is 1 sec by
default, but it can be increased using the system parameter
sale_order_product_picker.delay and the number of seconds to wait.

This relieves the server of unnecessary calls, since everything will be
written at the same time. And the user will not have the feeling of
slowness when writing on 3G networks.

CarlosRoca13 commented:
Added changes to make the button +1 faster. With these changes, what we do is to store the changes to be made while pressing the +1 button. Each time the button is pressed, the timer that will execute the writing of the lines is reset. 

This relieves the server of unnecessary calls, since everything will be written at the same time. And the user will not have the feeling of slowness when writing on 3G networks.

Here is a sample:
![Example picker16 0](https://github.com/OCA/sale-workflow/assets/35952655/c5362208-87eb-4a02-9136-49469d191435)

As you can see in the example, the time to write the pending lines is 1 sec, but it can be increased using the system parameter `sale_order_product_picker.delay` and the number of seconds to wait.

CarlosRoca13 commented:
@yajo I believe this should resolve this issue :smile: https://github.com/OCA/sale-workflow/issues/2842

sergio-teruel commented:
@CarlosRoca13 Can you put this PR as draft until september?

CarlosRoca13 commented:
This PR is ready for review

yajo reviewed (approved):
Thanks!

pedrobaeza commented:
/ocabot merge minor
